### DISCUSSION TITLE
Team D - Course 2

### DISCUSSION THREAD

**Discussion by 1hk-star** [lablup/contribution-academy-2021#29] [Show and tell]
Q1. 아래 저장소의 maintainer 를 찾고, 그 근거를 찾을 수 있는 git 명령어를 올려주세요.

Q2. Backend.AI project에 Write할 수 있는 권한이 있는 John과 권한이 없는 Jane 이 있습니다.
이 때, Jane 이 구현한 기능이 main branch에 적용되기 위한 단계는 John 이 기능을 main branch에
merge 할 때와 어떤 다른 점이 있고, 그 이유는 무엇인지 두 문장 이내로 간단하게 적어주세요.

**Comment by youngjun0627** [comment]
이유찬
Q1. 저는 `git log --pretty="%an" | sort | uniq -c` 를 써서 개수 많은 사람을 찾는 방식으로 했습니다. 리눅스 명령어를 많이 사용했는데 혹시 다른 명령어도 있다면 알려주세요.
Backend.AI-manager : Joongi Kim
Backend.AI-agent : Joongi Kim
Backend.AI-common : Joongi Kim
Backend.AI-client-py : Joongi Kim
Backend.AI-webserver : Jeongkyu Shin
Backend.AI-webui : Jeongkyu Shin
Backend.AI-storage-proxy : Joongi Kim

Q2. write 권한이 있는 John의 경우에는 직접 작업을 진행할 수 있으나, 권한이 없는 Jane의 경우에는 개인 repo에 fork하여 작업을 진행하고 PR을 해야됩니다. 이때 PR description에 이슈넘버를 포함하여 작성하여야 됩니다.

다른 팀원 분들의 조언 감사히 받겠습니다. 피드백 부탁드려용

**Reply by youngjun0627** [reply]
`git shortlog -s` 도 똑같이 출력되는거 같아요

![image](https://user-images.githubusercontent.com/57058726/128905281-1aa0d145-c69d-4bc7-bf0e-2153a965a489.png)

**Comment by 1hk-star** [comment]
Q1. 이유찬님 코드 그대로 사용했습니다.
Backend.AI-manager : Joongi Kim
Backend.AI-agent : Joongi Kim
Backend.AI-common : Joongi Kim
Backend.AI-client-py : Joongi Kim
Backend.AI-webserver : Jeongkyu Shin
Backend.AI-webui : Jeongkyu Shin
Backend.AI-storage-proxy : Joongi Kim

Q2. 
John의 경우 원본 저장소에 새로운 브랜치를 만들어 기능을 개발 후 main 브랜치에 머지합니다. John은 원본 저장소를 fork한 이후 본인 저장소에서 새로운 브랜치를 만들어 기능을 개발 후 원본 저장소의 main 브랜치에 머지합니다. 새 브랜치에서의 작업이 관리에 용이합니다.

**Comment by JHyunB** [comment]
Q1. 
```bash
git log | grep Author | sort | uniq -c | sort -n -r
```
Backend.AI-manager : Joongi Kim
Backend.AI-agent : Joongi Kim
Backend.AI-common : Joongi Kim
Backend.AI-client-py : Joongi Kim
Backend.AI-webserver : Jeongkyu Shin
Backend.AI-webui : Jeongkyu Shin
Backend.AI-storage-proxy : Joongi Kim

리눅스 명령어를 잘 몰라 힌트를 보고 maintainer를 찾았습니다. 힌트를 보았기 때문에 각 옵션에 대해 찾아보았습니다.
1. grep을 통해 Author 패턴을 갖는 줄을 가져옵니다.
2. 줄 내용을 바탕으로 정렬합니다.
3. uniq는 중복된 내용의 행이 연속으로 있으면 하나만 남기며 -c 를 통해 같은라인이 몇번나오는지를 표시합니다.
4. 마지막 sort를 통해 기여수 순으로 정렬합니다. sort는 기본적으로 오름차순 정렬입니다. -n : 라인의 각 필드를 비교하는 대상을 숫자로 한정  -r : 역순으로 정렬


Q2.
John은 쓰기 권한이 있기 때문에 원본 저장소 내부에 브랜치를 생성하고 작업을 하지만, Jane은 쓰기권한이 없기에 포크를 한 후, 포크한 저장소에 브랜치를 생성하고 작업해야 합니다. 이 이유는 질문에 나와있듯이, 쓰기 권한 차이 때문이라고 생각합니다.

**Comment by hyeyeon-sun** [comment]
Q1. git shortlog -s -n 명령어를 사용했습니다. 각 옵션에 대한 설명은 다음과 같습니다.
 - s : --summary
Suppress commit description and provide a commit count summary only.
 - n : --numbered
Sort output according to the number of commits per author instead of author alphabetic order.

Backend.AI-manager : Joongi Kim
Backend.AI-agent : Joongi Kim
Backend.AI-common : Joongi Kim
Backend.AI-client-py : Joongi Kim
Backend.AI-webserver : Jeongkyu Shin
Backend.AI-webui : Jeongkyu Shin
Backend.AI-storage-proxy : Joongi Kim

Q2. John의 경우 project에 Write할 수 있는 권한이 있기에 원본 저장소에 브랜치를 만들어 기능을 구현하고 merge 하면 됩니다. 그러나 Jane의 경우 권한이 없기 때문에 개인 저장소로 프로젝트를 fork해와서 브랜치를 만들어 변경사항을 커밋한 뒤 원본 저장소에 PR을 올려서 원본 저장소 관리자의 승인을 받아야 합니다. 이 때에 PR의 description에 이슈 넘버를 포함하여 작성해야 합니다.

**Comment by 1hk-star** [comment]
위 내용들로 정리하겠습니다.  @lizable 
Q1. 
git log --pretty="%an" | sort | uniq -c
git shortlog -s
git log | grep Author | sort | uniq -c | sort -n -r
git shortlog -s -n
등 여러 명령어를 통해 maintainer를 찾을 수 있었고 레포별 maintainer는 아래와 같습니다.

Backend.AI-manager : Joongi Kim
Backend.AI-agent : Joongi Kim
Backend.AI-common : Joongi Kim
Backend.AI-client-py : Joongi Kim
Backend.AI-webserver : Jeongkyu Shin
Backend.AI-webui : Jeongkyu Shin
Backend.AI-storage-proxy : Joongi Kim

Q2.
John은 쓰기 권한이 있기 때문에 원본 저장소 내부에 브랜치를 생성하고 작업을 하지만, Jane은 쓰기권한이 없기에 포크를 한 후, 포크한 저장소에 브랜치를 생성하고 작업해야 합니다. 이 이유는 질문에 나와있듯이, 쓰기 권한 차이 때문이라고 생각합니다.
